### DISCUSSION TITLE
Support "start phased release"

### DISCUSSION THREAD

**Discussion by mathaeus** [fastlane/fastlane#18190] [Ideas]
Hi, maybe I'm just missing something but it seems that there still is no way of starting a phased release on iOS.

The last discussion related to this topic I can find is this one https://github.com/fastlane/fastlane/issues/16519 which was closed without a solution. And I also checked the source code, there still doesn't seem to be anything - we only have the `release_to_all_users` method to **_complete_** a phased release, see https://github.com/fastlane/fastlane/pull/13941, but none to actually **_start_** it 🤔. 

Would the suggested solution from https://github.com/fastlane/fastlane/issues/16519 make sense? Looks straight-forward to me.

cc @max-ott

**Comment by dinsen** [comment]
Hi @mathaeus 
Have you seen this:

```ruby
Spaceship::ConnectAPI::AppStoreVersion.create_app_store_version_phased_release()
```
https://github.com/fastlane/fastlane/blob/master/spaceship/lib/spaceship/connect_api/models/app_store_version.rb#L161

It require you use the `API Key`:
https://docs.fastlane.tools/app-store-connect-api/

**Reply by mathaeus** [reply]
@Dinsen as far as I know this one is responsible for **_creating_** the version and set the "release type" to phased release, but not for **_starting_** the phased release itself.

I haven't gone through all of the source code but I assume that method is called if you pass `phased_release: true` in `deliver` (which we do).

e.g. we use this for preparing the version on the store:
```
    deliver(
      submit_for_review: false,
      automatic_release: false,
      phased_release: true,
      skip_screenshots: true,
      skip_metadata: false,
      skip_binary_upload: true,
      ...
    )
```

**Comment by mathaeus** [comment]
@TragedyStar did you ever find a proper solution for this? (I can't comment on your issue from back then since it's locked)

**Comment by mathaeus** [comment]
@joshdholtz (sorry for dragging you into this), is there any plan to move to the official App Store Connect API for managing releases?
i.e. https://developer.apple.com/documentation/appstoreconnectapi/app_store_version_phased_releases

The current implementations in `tunes_client` use the old ITC API from what I can see. So would it still make sense to add something like the following?

```
def start_phased_release!(app_id, version)
  ...
  r = request(:post) do |req|
    req.url("ra/apps/#{app_id}/versions/#{version}/phasedRelease/state/STARTED")
    req.headers['Content-Type'] = 'application/json'
    req.body = app_id.to_s
  end
  ...
end
```

**Reply by max-ott** [reply]
@mathaeus Fastlane / Spaceship is still open source. Feel free to open a pull request to add this using the new API.

**Reply by mathaeus** [reply]
@max-ott I was digging a bit more into it now.

From what I can see the new API is actually already implemented 😅

https://github.com/fastlane/fastlane/pull/17165

In order to start a phased release I would expect I can do the following:

```
version = app.get_pending_release_app_store_version(platform: Spaceship::ConnectAPI::Platform::IOS)
version.fetch_app_store_version_phased_release.resume
```

Note: 
- Here we try `get_pending_release_app_store_version` and not `get_live_app_store_version`, since well, the app is not live yet. 
- `resume` internally sets the state to `ACTIVE` which I assume is the proper state to start a phased release. (unfortunately the Apple docs are not really descriptive...)

----

Actually a colleague of mine tried that before (I now understand how he came up with the code). Yet, unfortunately we were not able to **_start_** the phased release like this:

```
Spaceship::UnexpectedResponse: [!] An attribute value is not acceptable for the current resource state. - You cannot change the state of a phased release in the current version state.
  path/to/ruby/gems/fastlane-2.166.0/spaceship/lib/spaceship/connect_api/api_client.rb:193:in `handle_response'
  path/to/ruby/gems/fastlane-2.166.0/spaceship/lib/spaceship/connect_api/api_client.rb:135:in `patch'
  path/to/ruby/gems/fastlane-2.166.0/spaceship/lib/spaceship/connect_api/tunes/tunes.rb:730:in `patch_app_store_version_phased_release'
  /Users/distiller/.rubies/ruby-2.7.2/lib/ruby/2.7.0/forwardable.rb:235:in `patch_app_store_version_phased_release'
  path/to/ruby/gems/fastlane-2.166.0/spaceship/lib/spaceship/connect_api/models/app_store_version_phased_release.rb:51:in `update'
  path/to/ruby/gems/fastlane-2.166.0/spaceship/lib/spaceship/connect_api/models/app_store_version_phased_release.rb:39:in `resume'
```

@polmum since you were the creator of the PR to add this. Were you able to start a phased release like this?

**Reply by polmum** [reply]
Hello!
You can't technically start a phased release. As you mentioned previously, you set the "release type" to phased release with deliver and when the app is released it automatically starts the phased release.
What you appear to want to do is manually release an app version that is in the PENDING_DEVELOPER_RELEASE state. I use the automatic_release configuration so I don't have this issue, but after checking it appears that manually releasing a version is not yet supported by the AppStore Connect API.
Luckily, someone seems to have found a [workaround](https://github.com/fastlane/fastlane/issues/17283#issuecomment-711865915). Basically, once your app is in the PENDING_DEVELOPER_RELEASE state, you schedule it for an automatic release in a past date, and after a couple of minutes the phased release should start.

**Reply by mathaeus** [reply]
@polmum thanks for linking that workaround and giving your insights/experience! :)

Yes, exactly. I'm trying to "manually" start the phased release. Yeah it seems it's not supported then.
That's why I said "`resume` internally sets the state to `ACTIVE` which I assume is the proper state to start a phased release." but I guess that's really only for resuming after pausing it before then.

I will try out that workaround and see if it works!

**Reply by mathaeus** [reply]
The workaround doesn't seem to work for me :/ 

It does set the scheduled date properly, but then nothing happens. I tried both:

### Setting the scheduled date in the past (last full hour) - took the screenshot after 8
<img width="490" alt="Screenshot 2021-02-22 at 08 41 28" src="https://user-images.githubusercontent.com/5389635/108698322-4c93b680-7504-11eb-9862-549442cdbd07.png">

### Setting the scheduled date in the future (next full hour) - took the screenshot at 09:37 local time
<img width="423" alt="Screenshot 2021-02-22 at 09 37 44" src="https://user-images.githubusercontent.com/5389635/108698389-603f1d00-7504-11eb-8e4b-4e50d2f343bb.png">


 ℹ️ I was waiting for more than an hour in both cases.


@dcacenabes does the workaround **_still_** work for you? Last time [you confirmed 24 days ago](https://github.com/fastlane/fastlane/issues/17283#issuecomment-769683371)?  :)

**Reply by dcacenabes** [reply]
Hi @mathaeus 
Last week it was still working, but today we are doing a new release and it is not working. We are experimenting the exact same thing as you, waiting for a couple of hours, date in the past and in the future. No luck.

**Reply by mathaeus** [reply]
Thanks so much for the quick answer @dcacenabes.

Well, then I'm glad I at least tried it correctly, but yeah too bad it doesn't work anymore.

**Reply by lingzlu** [reply]
Confirming that the workaround no longer works, waited 3 hours and still not releasing. :(

**Reply by ainame** [reply]
I'm crying here that I implemented the workaround a few weeks ago and rolled it out for the last Monday's release and turned out my change won't work without being able to work even once😭 but thank you for sharing tips.

I will try something different for next week (we are doing weekly releasing)

**Reply by dcacenabes** [reply]
@ainame sorry to hear that. If I may ask, what are you planning to test?

**Reply by ainame** [reply]
No plan yet. I will figure it out😂

**Reply by sergeykochanov** [reply]
It seems the only way is to go back to old API with 2FA. :(

**Reply by dcacenabes** [reply]
@mathaeus I am starting to think that the problem was not actually with the workaround itself (changing from manual to scheduled release), but rather with some change in automatic releases in Apple's side.
Currently we have an app scheduled to automatically be released today at 9.00AM GMT. We did not change from Manual to Scheduled release once the app was already approved. This release's release date was specified when we submitted for review and it has always been scheduled. We are one hour past the scheduled time and the app is still Pending Developer Release.
<img width="472" alt="Screenshot 2021-03-15 at 10 55 02" src="https://user-images.githubusercontent.com/3796899/111135403-e5a27400-857c-11eb-82f3-633af0be006f.png">

**Reply by mathaeus** [reply]
@dcacenabes thanks for the update.
I haven't scheduled any app release in the regular way like this, but that would make sense if you ask me :)

So yes, "let's hope" it's a general bug on Apple's side (which they hopefully fix soon), so that we could reapply the workaround as long as there's no official API for this in the first place.

**Reply by sergeykochanov** [reply]
@dcacenabes Didn't you submit a ticket to Apple https://feedbackassistant.apple.com/?

**Reply by dcacenabes** [reply]
> @dcacenabes Didn't you submit a ticket to Apple https://feedbackassistant.apple.com/?

Yes I did -> FB9043418

**Reply by ainame** [reply]
Hey! Our app was approved last night and this morning tried to run this code snippet, which is exactly the same as suggested on this thread, on `pry` command manually. It is the same one that failed last week, too. 

```ruby
            beginning_of_hour_to_release = Time.now.strftime('%Y-%m-%dT%H:00%:z')
            app_store_version.update(
              attributes: {
                earliest_release_date: beginning_of_hour_to_release,
                release_type: Spaceship::ConnectAPI::AppStoreVersion::ReleaseType::SCHEDULED
              }
            )
```

However, it worked fine this week. I'm almost sure it was the bug and fixed😂 

I'm gonna share some logs on `pry` (not interesting though)

<details>

```
[5] pry(#<Fastlane::FastFile>)> beginning_of_hour_to_release = Time.now.strftime('%Y-%m-%dT%H:00%:z')
=> "2021-03-17T09:00+00:00"
[6] pry(#<Fastlane::FastFile>)> app_store_version = Spaceship::ConnectAPI::App.find('bundle id').get_pending_release_app_store_version
=> #<Spaceship::ConnectAPI::AppStoreVersion:0x00007f9bc1b52070
 @app_store_state="PENDING_DEVELOPER_RELEASE",
 @app_store_version_submission=#<Spaceship::ConnectAPI::AppStoreVersionSubmission:0x00007f9bc1b50f90 @id="cec6ff3e-e4b9-4094-befc-f7f9d67e7993">,
 @build=
  #<Spaceship::ConnectAPI::Build:0x00007f9bc1b51bc0
   @expiration_date="2021-06-10T08:31:07-07:00",
   @expired=false,
   @icon_asset_token={"templateUrl"=>"xxxxxxxxxxxxxxxxxx", "width"=>167, "height"=>167},
   @id="xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx",
   @min_os_version="12.0",
   @processing_state="VALID",
   @uploaded_date="2021-03-12T08:31:07-08:00",
   @uses_non_exempt_encryption=false,
   @version="4210">,
 @copyright="xxxxxxxxxxxx",
 @created_date="2021-03-10T05:08:58-08:00",
 @downloadable=true,
 @earliest_release_date=nil,
 @id="xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx",
 @platform="IOS",
 @release_type="MANUAL",
 @uses_idfa=false,
 @version_string="10.29.0">
[7] pry(#<Fastlane::FastFile>)>             app_store_version.update(
              attributes: {
                earliest_release_date: beginning_of_hour_to_release,
                release_type: Spaceship::ConnectAPI::AppStoreVersion::ReleaseType::SCHEDULED
              }
[7] pry(#<Fastlane::FastFile>)> app_store_version.update(
  attributes: {
    earliest_release_date: beginning_of_hour_to_release,
    release_type: Spaceship::ConnectAPI::AppStoreVersion::ReleaseType::SCHEDULED
  }
=> #<Spaceship::ConnectAPI::AppStoreVersion:0x00007f9bf1c4a150
 @app_store_state="PENDING_DEVELOPER_RELEASE",
 @copyright="xxxxxxxxxxxx",
 @created_date="2021-03-10T05:08:58-08:00",
 @downloadable=true,
 @earliest_release_date="2021-03-17T02:00:00-07:00",
 @id="xxxxxxxxxxxxxxx",
 @platform="IOS",
 @release_type="SCHEDULED",
 @uses_idfa=false,
 @version_string="10.29.0">
```

And then the app was released a few mins later...

</details>

**Reply by ainame** [reply]
I will probably add polling logic to check when the app store version becomes "Ready to Sale" state. For example, like below.

```ruby
beginning_of_hour_to_release = Time.now.strftime('%Y-%m-%dT%H:00%:z')
app_store_version.update(
  attributes: {
    earliest_release_date: beginning_of_hour_to_release,
    elease_type: Spaceship::ConnectAPI::AppStoreVersion::ReleaseType::SCHEDULED
  }
)

loop do
  # It would be good to add some kind of timeout here, too 
  version = app.get_pending_release_app_store_version
  break unless version 
  sleep(60)
end

UI.message("The new version was successfully released! 🎉")
```

This way we don't have to worry about the rest of the release flow that might fail without the new version released.

**Reply by mathaeus** [reply]
Thanks for the update @ainame! Sounds promising then that the workaround would work again.

@dcacenabes did you receive any feedback on your "feedback ticket" yet? :)

**Reply by dcacenabes** [reply]
Thanks @ainame ! Next Monday I will give it a shot too!
@mathaeus  Regarding my feedback ticket, no updates yet :(

I think that if more people confirm this is working once again we should add the workaround available tag again to https://github.com/fastlane/fastlane/issues/17283#issuecomment-711865915

**Reply by ainame** [reply]
I've tried the above polling logic this week and then it took us about 5 mins to get the new version released. It seems to work well now 🎉

**Reply by joshdholtz** [reply]
Hello 👋 This is a long thread that I apparently missed 🤦 What is the TL;DR here? Is there something that we can do in _fastlane_ to make this easier? 😇

**Reply by dcacenabes** [reply]
Hi @joshdholtz!

The TL;DR would be that App Store Connect API does not support releasing an app immediately, a call to the method "create_app_store_version_release_request" [will produce an error](https://github.com/fastlane/fastlane/issues/17283).
[I found a workaround](https://github.com/fastlane/fastlane/issues/17283#issuecomment-711865915) by setting the release to scheduled to some minutes in the past. It usually becomes published just after a few minutes, but the downside is that there is a delay between your call and the actual release.

~~The TL;DR would be that App Store Connect API does not support releasing an app if it has a phased release, a call to the method "create_app_store_version_release_request" [will produce an error](https://github.com/fastlane/fastlane/issues/17283).
[I found a workaround](https://github.com/fastlane/fastlane/issues/17283#issuecomment-711865915) by setting the release to scheduled to some minutes in the past. It usually becomes published just after a few minutes.~~

I am not sure if it makes sense to include this in Fastlane somehow, but for sure it would make sense that create_app_store_version_release_request checks if there is a phased released and then fails more gracefully, maybe letting the user that there is a workaround available?

**Reply by ainame** [reply]
The thing is that it's not only about "Phased release" but also about "Manual immediate release via fastlane" without "phased release".🙂 I want the equivalent of `app_store_version.create_app_store_version_release_request` that works with JWT session. Even more, I (and some people, perhaps) want it to be an action in fastlane to not duplicate workaround code in every project😄 

If we can agree with adding that action, I can submit a PR to turn the workaround of @dcacenabes idea into action 👍  
https://github.com/fastlane/fastlane/discussions/18190#discussioncomment-493265

cc: https://github.com/fastlane/fastlane/discussions/17246

**Reply by dcacenabes** [reply]
> The thing is that it's not only about "Phased release" but also about "Manual immediate release via fastlane" without "phased release".🙂 I want the equivalent of `app_store_version.create_app_store_version_release_request` that works with JWT session. Even more, I (and some people) want it to be an action in fastlane to not duplicate workaround code in every project😄
> 
> If we can agree with adding that action, I can submit a PR to turn the workaround of @dcacenabes idea into action 👍
> [#18190 (reply in thread)](https://github.com/fastlane/fastlane/discussions/18190#discussioncomment-493265)
> 
> cc: #17246

Sorry because I think I misunderstood previous messages. I thought that app_store_version.create_app_store_version_release_request was working if the release was not phased.

I am going to update my previous message for clarification.

**Reply by dcacenabes** [reply]
Update on Apple's feedback FB9043418: After sharing information about the app version and dates when this happened, they confirmed that the issue is resolved (although we already knew that), so they indirectly confirmed that the automatic release not working was an issue in their end.

**Comment by kambala-decapitator** [comment]
I use this to release app on AppStore: https://github.com/fastlane/fastlane/issues/17140#issuecomment-686493328

Phased release is turned on beforehand in `deliver` call.

**Reply by LancerTZ** [reply]
Really??? As far as I know, there is no direct solution to start a phased release.

Maybe you only turned on the phased release state?

**Reply by kambala-decapitator** [reply]
@TragedyStar I use it to **start release**, phased release state is set in `deliver` when I submit for review. That's not a built-in command indeed: it uses `spaceship`, as you can see from the link. Initially figured it out in https://github.com/fastlane/fastlane/discussions/16994.

**Reply by LancerTZ** [reply]
Amazing! I see many people still using this to workaround

```
date = Time.now.strftime('%Y-%m-%dT%H:00%:z')
version.update(attributes: {
  earliest_release_date: date,
  release_type: Spaceship::ConnectAPI::AppStoreVersion::ReleaseType::SCHEDULED
})
```

I will have a deep look in spaceship source code and have a re-try. I tried this last year, but I remember it doesn't work.

```
require ‘spaceship’
Spaceship::ConnectAPI.login(“your@email.com”, team_name: “Your Team Name”)
app = Spaceship::ConnectAPI::App.find(“com.bundle.id”)
version = app. get_pending_release_app_store_version(platform: Spaceship::ConnectAPI::Platform::IOS)
Spaceship::ConnectAPI.post_app_store_version_release_request(app_store_version_id: version.id)
```

**Reply by kambala-decapitator** [reply]
here's an excerpt from my project:
```ruby
teamID = ...
bundleID = ...

Spaceship::ConnectAPI.login(
  use_portal: false,
  tunes_team_id: teamID,
)

app = Spaceship::ConnectAPI::App.find(bundleID)
if app.nil?
  UI.important "not found in team #{teamID} on ASC!"
  next
end

version = app.get_pending_release_app_store_version(platform: Spaceship::ConnectAPI::Platform::IOS)
if version.nil?
  UI.important "not pending release!"
  next
end

Spaceship::ConnectAPI.post_app_store_version_release_request(app_store_version_id: version.id)
UI.success "released v#{version.version_string}"
```

**Comment by Johennes** [comment]
If anyone else came here from a Google search and is confused: If you don't want to or cannot specify the `phased_release` flag in `deliver`, you can use something like this to toggle the setting:

```
lane :update_phased_release do |options|
  phased_release = options[:version].fetch_app_store_version_phased_release rescue nil

  if options[:activate]
    unless phased_release
      options[:version].create_app_store_version_phased_release(attributes: {
        phasedReleaseState: Spaceship::ConnectAPI::AppStoreVersionPhasedRelease::PhasedReleaseState::INACTIVE
      })
    end
  else
    if phased_release
      phased_release.delete!
    end
  end
end
```

Calling that lane via e.g.

```
app = Spaceship::ConnectAPI::App.find(...)
version = app.get_pending_release_app_store_version(platform: Spaceship::ConnectAPI::Platform::IOS)
update_phased_release(version: version, activate: true)
```

is the equivalent of making this setting in AppStoreConnect:

![Screenshot 2021-06-01 at 11 07 46](https://user-images.githubusercontent.com/1137962/120320035-fcd24180-c2e1-11eb-9129-4bdf2282ed59.png)

Once you've changed the setting to what you want it to be, you can release the version like normally, e.g. using the "scheduled release in the past" trick mentioned in one of the comments above:

```
date = Time.now.strftime('%Y-%m-%dT%H:00%:z')
version.update(attributes: {
  earliest_release_date: date,
  release_type: Spaceship::ConnectAPI::AppStoreVersion::ReleaseType::SCHEDULED
})
```

It will then take a few minutes until the automatic release happens.

So ultimately if you need a "start phased release" functionality, all you have to do is combine the three snippets above into a dedicated lane.
